File: examples/clj/test/com/rpl/agent/basic/streaming_agent_test.clj
(ns com.rpl.agent.basic.streaming-agent-test
  (:require
   [clojure.test :refer [deftest testing is]]
   [com.rpl.agent-o-rama :as aor]
   [com.rpl.rama :as rama]
   [com.rpl.rama.test :as rtest]
   [com.rpl.agent.basic.streaming-agent :refer [StreamingAgentModule]]))

(deftest streaming-agent-test
  (System/gc)
  (testing "StreamingAgent example produces expected streaming behavior"
    (with-open [ipc (rtest/create-ipc)]
      (rtest/launch-module! ipc StreamingAgentModule {:tasks 1 :threads 1})

      (let [manager (aor/agent-manager ipc
                                       (rama/get-module-name
                                        StreamingAgentModule))
            agent   (aor/agent-client manager "StreamingAgent")]

        (testing "streams progress chunks and returns final result"
          (let [invoke (aor/agent-initiate agent {:data-size 20 :chunk-size 5})
                chunks-received (atom [])]

            ;; Subscribe to streaming chunks
            (aor/agent-stream
             agent
             invoke
             "process-data"
             (fn [all-chunks new-chunks reset? complete?]
               (doseq [chunk new-chunks]
                 (swap! chunks-received conj chunk))))

            ;; Get final result
            (let [result (aor/agent-result agent invoke)]
              ;; Verify final result structure
              (is (= "data-processing" (:action result)))
              (is (= 20 (:total-items result)))
              (is (= 4 (:total-chunks result)))
              (is (= 5 (:chunk-size result)))
              (is (number? (:completed-at result)))

              ;; Verify streaming chunks were received
              (is (= 4 (count @chunks-received)))

              ;; Verify chunk structure and progress
              (let [chunks @chunks-received]
                (is (every? #(contains? % :chunk-number) chunks))
                (is (every? #(contains? % :items-processed) chunks))
                (is (every? #(contains? % :progress) chunks))
                (is (every? #(contains? % :items) chunks))

                ;; Verify progress increases
                (is (= [0.25 0.5 0.75 1.0] (map :progress chunks)))

                ;; Verify chunk numbers
                (is (= [0 1 2 3] (map :chunk-number chunks)))

                ;; Verify items processed per chunk
                (is (= [5 5 5 5] (map :items-processed chunks)))))))))))

File: examples/java/basic/src/test/java/com/rpl/agent/basic/StreamingAgentTest.java
package com.rpl.agent.basic;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import com.rpl.agentorama.AgentClient;
import com.rpl.agentorama.AgentInvoke;
import com.rpl.agentorama.AgentManager;
import com.rpl.rama.test.InProcessCluster;
import com.rpl.rama.test.LaunchConfig;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.Test;

/**
 * Test class for StreamingAgent demonstrating streaming data patterns.
 *
 * <p>This test demonstrates:
 *
 * <ul>
 *   <li>streamChunk: Emitting streaming data from agent nodes
 *   <li>agent.stream: Subscribing to streaming data from specific nodes
 *   <li>Real-time data flow with incremental results
 *   <li>Streaming completion and callbacks
 * </ul>
 */
public class StreamingAgentTest {

  @Test
  public void testStreamingAgentBasicFunctionality() throws Exception {
    // Tests basic streaming functionality
    try (InProcessCluster ipc = InProcessCluster.create()) {
      // Deploy the agent module
      StreamingAgent.StreamingAgentModule module = new StreamingAgent.StreamingAgentModule();
      ipc.launchModule(module, new LaunchConfig(1, 1));

      // Get agent manager and client
      String moduleName = module.getModuleName();
      AgentManager manager = AgentManager.create(ipc, moduleName);
      AgentClient agent = manager.getAgentClient("StreamingAgent");

      // Start agent execution with small numbers for testing
      Map<String, Object> request = new HashMap<>();
      request.put("dataSize", 10); // 10 total items
      request.put("chunkSize", 5); // chunk size 5
      AgentInvoke invoke = agent.initiate(request);

      // Track chunks received via streaming
      AtomicInteger chunksReceived = new AtomicInteger(0);
      List<Map<String, Object>> receivedChunks = new ArrayList<>();

      // Subscribe to streaming chunks
      agent.stream(
          invoke,
          "process-data",
          (allChunks, newChunks, reset, complete) -> {
            for (Object chunkObj : newChunks) {
              @SuppressWarnings("unchecked")
              Map<String, Object> chunk = (Map<String, Object>) chunkObj;
              receivedChunks.add(chunk);
              chunksReceived.incrementAndGet();
            }
          });

      // Get final result
      @SuppressWarnings("unchecked")
      Map<String, Object> result = (Map<String, Object>) agent.result(invoke);

      // Verify streaming chunks were emitted
      assertTrue("Should have received streaming chunks", chunksReceived.get() > 0);
      assertTrue("Should have received chunks", receivedChunks.size() > 0);

      // Verify final result
      assertNotNull("Final result should not be null", result);
      assertEquals("Should process 10 items", 10, (int) result.get("totalItems"));
      assertEquals("Should have chunk size 5", 5, (int) result.get("chunkSize"));
      assertTrue("Should have total chunks", (Integer) result.get("totalChunks") > 0);
    }
  }
}
